### PR TITLE
feat: migration to nuxt3

### PR DESCRIPTION
**Pull request by antfu**
## Progress

- [x] Nuxt 2 -> Nuxt 3
- [x] Yarn v1 -> pnpm
- [x] Axios -> ohmyfetch
- [x] JS -> TS
- [x] Rewrite all components to `<script setup>`
- [x] `vaildate` to route middleware
- [x] https://github.com/nuxt/framework/issues/3242
- [x] Investigate 404 on initial load
- [x] `useState` over `Pinia`
- [x] Performance measurement

Preview: https://hackernews-git-nuxt3-nuxt-js.vercel.app/

### CODE REVIEW

**Comment by pi0** [comment]
> Yarn v1 -> pnpm

pnpm is still not in a state that we can use in at least official repo and examples and advocate yet. There are issues like (https://github.com/unjs/nitro/issues/106) but I guess more are still there.

**Comment by antfu** [comment]
Do you want to switch to Yarn v3, or?

File: layouts/default.vue
Comment (antfu) [inline layouts/default.vue:78]:
```suggestion
    &.router-link-active, &.nuxt-link-active {
```

File: middleware/feed.ts
Comment (danielroe) [inline middleware/feed.ts:3]: for now, can we redirect to a safe path instead?

Comment (antfu) [reply middleware/feed.ts]: Oh indeed! I made the changes yesterday but failed to push and didn't notice 😅

File: composables/store.ts
Comment (Atinux) [inline composables/store.ts:2]: I think it can be removed since we have the auto-imports

**Comment by antfu** [comment]
Instead of merging, I think we could just switch to the default branch to `nuxt3` (I already made the `nuxt2` branch as a pointer). This PR serves the purpose of reviewing.

**Comment by Atinux** [comment]
I like this idea

File: README.md
Comment (antfu) [inline README.md:17]: I guess the SPA mode is not that cool in Nuxt 3 anymore. Since we should many hosting services out-of-box, I am thinking maybe here we could deploy to all of them at the same time and provide the previews like:

- nuxt3-hn.vercel.app
- nuxt3-hn.netlify.app
- nuxt3-hn.cloudflare.app
- ....

Comment (anburocky3) [reply README.md:17]: 💡 Great idea.

Comment (Atinux) [reply README.md:17]: I like the idea. SPA is still cool since it shows that you simply need to disable one line but I am fine removing it for Nuxt 3 examples and instead showcase all the providers it supports.

Comment (pi0) [reply README.md:17]: Nitro preset deploys are already deployed on (https://github.com/unjs/nitro-deploys). It is a huge double work for maintenance if we do it with each example.

Comment (antfu) [reply README.md:17]: Oh, that makes sense then! Do you think we should keep Vercel only, or use the old format (Vercel for SSR and Netlify for SPA?)

Comment (Atinux) [reply README.md:17]: Yes we can keep using these two, forgot about this double maintenance

**Comment by Atinux** [comment]
Shall we wait for https://github.com/nuxt/framework/issues/3242 to be fixed before merging?

We can anyway announce it after RC to keep momentum 😊

**Comment by Atinux** [comment]
I let @pi0 decide about the package manager

**Review by pi0 — DISMISSED** [review]
Please do not merge it yet!! I need to review the changes following nuxt3.

**Comment by Atinux** [comment]
Interesting: https://hackernews-17mm8jqll-nuxt-js.vercel.app/

It show 404 page and redirects, then if you refresh the page is good.

File: middleware/feed.ts
Comment (Atinux) [inline middleware/feed.ts:2]: Where is defined validFeeds?

I will redirect directly to validFeeds[0] + '/1' to avoid multiple redirects

Comment (antfu) [reply middleware/feed.ts:2]: https://github.com/nuxt/hackernews/blob/e160b93f1f1be4825f9690936b8b948e618c4f1e/composables/api.ts#L9

and that's a good point!

**Comment by Atinux** [comment]
Do we need to use Pinia or can we simply use `useState` for that?

**Comment by Atinux** [comment]
What are the tasks left @antfu ?

I still see the 404 page on https://hackernews-kv91rg0gq-nuxt-js.vercel.app at first

**Comment by antfu** [comment]
> What are the tasks left 

I have updated the description. Thanks for the reminder!

> I still see the 404 page on

Yeah I am still investigating it

**Comment by antfu** [comment]
It's ready :)

image

**Comment by Atinux** [comment]
Seems like we have some issues with transitions, do you mind checking what is creating this flash?

https://user-images.githubusercontent.com/904724/165923006-7f0b3cc5-d08b-4737-b21e-a35bcaffa911.MOV

**Comment by Atinux** [comment]
Any update regarding my last comment @antfu ?

**Comment by antfu** [comment]
I have temporarily removed the transition as I think there is a bug in Nuxt 3 that that page instance is not preserved on sub-routes. Will investigate.

**Comment by antfu** [comment]
I have reworked the lazying loading part. It should preforms better now :)

File: components/Comment.vue
Comment (Atinux) [inline components/Comment.vue:21]:
```suggestion
      <NuxtLink :to="'/user/' + comment.user">
```

File: components/Item.vue
Comment (Atinux) [inline components/Item.vue:18]:
```suggestion
        <NuxtLink :to="'/item/' + item.id">{{ item.title }}</NuxtLink>
```

File: components/Item.vue
Comment (Atinux) [inline components/Item.vue:25]:
```suggestion
        <NuxtLink :to="'/user/' + item.user">{{ item.user }}</NuxtLink>
```

File: components/ItemListNav.vue
Comment (Atinux) [inline components/ItemListNav.vue:13]:
```suggestion
    <NuxtLink v-if="page > 1" :to="`/${feed}/${page - 1}`">
```

File: components/ItemListNav.vue
Comment (Atinux) [inline components/ItemListNav.vue:15]:
```suggestion
    </NuxtLink>
```

File: components/ItemListNav.vue
Comment (Atinux) [inline components/ItemListNav.vue:18]:
```suggestion
    <NuxtLink v-if="hasMore" :to="`/${feed}/${page + 1}`">
```

File: components/ItemListNav.vue
Comment (Atinux) [inline components/ItemListNav.vue:20]:
```suggestion
    </NuxtLink>
```

File: layouts/default.vue
Comment (Atinux) [inline layouts/default.vue:21]:
```suggestion
        <NuxtLink to="/" exact>
```

File: layouts/default.vue
Comment (Atinux) [inline layouts/default.vue:23]:
```suggestion
        </NuxtLink>
```

File: layouts/default.vue
Comment (Atinux) [inline layouts/default.vue:26]:
```suggestion
        </NuxtLink>
```

File: pages/item/[id].vue
Comment (Atinux) [inline pages/item/[id].vue:33]:
```suggestion
          <NuxtLink :to="'/user/' + item.user">
```

File: pages/item/[id].vue
Comment (Atinux) [inline pages/item/[id].vue:35]:
```suggestion
          </NuxtLink>
```

**Comment by Atinux** [comment]
Added a review for using `NuxtLink` instead of `RouterLink`

File: app.vue
Comment (Atinux) [inline app.vue:5]: Can we generate another image based on your latest design updates @antfu ?

Comment (antfu) [reply app.vue:5]: Frame 1 (1)

Made a simple design of it, WDYT?

Comment (Atinux) [reply app.vue:5]: Awesome!

Comment (antfu) [reply app.vue]: Thanks, updated!

**Review by Atinux — APPROVED** [review]

File: components/ItemListNav.vue
Comment (Atinux) [inline components/ItemListNav.vue:16]:
```suggestion
    <span v-else class="disabled">&lt; prev</span>
```

**Comment by Atinux** [comment]
Thank you @antfu and sorry for the long time reviewing it.

So far we won't push the marketing of it until we push the PWA support into Nuxt 3 with Pooya 😊

**Pull request merged**
